### PR TITLE
DEVPROD-2299 Add CI task for go vulnerabilities

### PR DESCRIPTION
**Pull request by ZackarySantana**
DEVPROD-2299

### Description
This adds a go task for go vulnerabilities. It adds a script for users to run locally if they would like and a make command to help with it

### Testing
Manual and PR patches

### CODE REVIEW

File: makefile [hunk 1]
```diff
@@ -361,6 +361,9 @@ gqlgen:
 	$(gobin) run github.com/99designs/gqlgen generate
 	$(gobin) run cmd/gqlgen/generate_secret_fields.go
 
+govul-install:
```
Comment (bsamek) [inline makefile:364]: For discoverability (some shells will autocomplete `make` targets, and people may read the `makefile` themselves), I would suggest adding a target to run `scripts/check-go-vulnerabilities.sh`, and calling that `make` target from `self-tests.yml`.

File: scripts/check-go-vulnerabiltiies.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#!/usr/bin/env bash
+
+# This script checks for go vulnerabilities.
+
+govul="govulncheck"
+
+# Test if govulncheck is installed as a binary or a go module.
+if command -v govul &> /dev/null; then
+    # If the user has installed swaggo as a binary, use the binary.
+    swaggo=$(command -v govul)
```
Comment (bsamek) [inline scripts/check-go-vulnerabiltiies.sh:11]: Is the `swaggo` here from a copy/paste maybe?

Comment (bsamek) [inline scripts/check-go-vulnerabiltiies.sh:11]: Is `govul` ever a binary?

**Review by bsamek — CHANGES_REQUESTED** [review]

File: scripts/check-go-vulnerabiltiies.sh (hunk 2 above)
Comment (ZackarySantana) [reply scripts/check-go-vulnerabiltiies.sh:11]: Yup should be switched to govul!

Comment (ZackarySantana) [reply scripts/check-go-vulnerabiltiies.sh:11]: No, this should be either the variable or hardcoded. I'll make it hard coded for clarity

File: makefile (hunk 1 above)
Comment (ZackarySantana) [reply makefile:364]: I'm not convinced about adding a target for it since to run it locally, it would be the same as all the other verify scripts. The other verify scripts also do not have a target (only one does and it's instead a go command rather than a shell script). I think I should either add all the scripts to the makefile and rewrite the self-tests to use those targets or write none to keep consistency. I'm leaning towards not because it adds length to the makefile.

Alternatively I thought about using a wildcard syntax in the makefile since most follow a "verify-" and then adding only a couple commands but it wouldn't have great autocomplete then. What do you think?

Comment (bsamek) [reply makefile:364]: That (not adding to the makefile) sounds reasonable to me.

File: scripts/check-go-vulnerabilities.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
+
```
Comment (bsamek) [inline scripts/check-go-vulnerabilities.sh:1]: In case you didn't see, this script is working locally for me, but is failing in CI.

File: scripts/check-go-vulnerabilities.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+fi
+
+# If govulncheck is not installed, exit with an error.
+if ! command -v "$govul" &> /dev/null; then
+    echo "govulncheck is not installed."
+    exit 1
+fi
+
+result=$($govul ./...)
+if [ $? -eq 0 ]; then
```
Comment (bsamek) [inline scripts/check-go-vulnerabilities.sh:25]: It might be nice if this prints a message so that users know no vulnerabilities were found. This distinguishes it from the script crashing or exiting without a useful error message.

**Review by bsamek — CHANGES_REQUESTED** [review]

File: scripts/check-go-vulnerabilities.sh (hunk 3 above)
Comment (ZackarySantana) [reply scripts/check-go-vulnerabilities.sh:1]: Yeah... I've been trying to debug it but I'm really not sure why it's giving the error in CI. I tested with the same exact go version and according to the logs it has the error in stderr but then it says "no vulnerabilities found" in stdout. If you have some insight on the tool please lmk! If not no worries, I will continue to debug 🔨

File: scripts/check-go-vulnerabilities.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    govul=$(command -v govulncheck)
+elif [ -n "$GOPATH" ] && [ -f "$GOPATH/bin/govulncheck" ]; then
+    govul="$GOPATH/bin/govulncheck"
+elif [ -n "$GOBIN" ] && [ -f "$GOBIN/govulncheck" ]; then
+    govul="$GOBIN/govulncheck"
+fi
+
+# If govulncheck is not installed, exit with an error.
+if ! command -v "$govul" &> /dev/null; then
+    echo "govulncheck is not installed."
```
[the remark this answers is not in the document]
Comment (ZackarySantana) [reply scripts/check-go-vulnerabilities.sh:20]: I agree, changing it!

File: scripts/check-go-vulnerabilities.sh (hunk 4 above)
Comment (ZackarySantana) [reply scripts/check-go-vulnerabilities.sh:25]: I also changed it for the swaggo fmt file to be more consistent/helpful

File: scripts/check-go-vulnerabilities.sh (hunk 3 above)
Comment (bsamek) [reply scripts/check-go-vulnerabilities.sh:1]: I wonder if it is some weird shell thing. Maybe try `set -x` at the top of the script to verify that it's doing what you expect?

**Comment by ZackarySantana** [comment]
I figured it out, the go binary govulncheck was pulling from the path was an old version (1.7) which did not have the -compile flag on the 'go list' subcommand (combed through the go versions and the commands- had to go _far_ back).

I added an export line and it works as intended in CI now.

**Comment by ZackarySantana** [comment]
I'm not worried about the failing parsely and spruce tasks, this should have absolutely no effect on them

**Review by bsamek — APPROVED** [review]
